### ISSUE TITLE
Multidimentional arrays generate an unexpected warning

### ISSUE THREAD

**Issue by earnol** [llvm/llvm-project#159603]
Since The commit https://github.com/llvm/llvm-project/commit/b24769855d97697de08e2296a548c033f193caf4#diff-778d8768bc77e1375d9a34c93e3885c1a986c00baa2435489742f270aede51eb the following code:
```
#include <stdio.h>
int afunc(unsigned long * v)
{
    return v != NULL;
}
unsigned long mdarr[5][6];
unsigned long sdarr[30];
int main()
{
  afunc(sdarr);
  afunc(mdarr);
  printf("mdarr = %zu \n", sizeof(mdarr));
  printf("sdarr = %zu \n", sizeof(sdarr));
  return 0;
}
```
generate warning: 
`error: incompatible pointer types passing 'unsigned long[5][6]' to parameter of type 'unsigned long *' [-Wincompatible-pointer-types]`.
Please see [godbolt example](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:12,endLineNumber:17,positionColumn:12,positionLineNumber:17,selectionStartColumn:12,selectionStartLineNumber:17,startColumn:12,startLineNumber:17),source:'%23include+%3Cstdio.h%3E%0A%0Aint+afunc(unsigned+long+*+v)%0A%7B%0A++++return+v+!!%3D+NULL%3B%0A%7D%0A%0Aunsigned+long+mdarr%5B5%5D%5B6%5D%3B%0Aunsigned+long+sdarr%5B30%5D%3B%0A%0Aint+main()%0A%7B%0A++afunc(sdarr)%3B%0A++afunc(mdarr)%3B%0A++printf(%22mdarr+%3D+%25zu+%5Cn%22,+sizeof(mdarr))%3B%0A++printf(%22sdarr+%3D+%25zu+%5Cn%22,+sizeof(sdarr))%3B%0A++return+0%3B%0A%7D%0A%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),header:(),k:34.71001224185383,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cclang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-O0+-Wincompatible-pointer-types',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),header:(),k:25.7599630930296,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+clang+20.1.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),k:39.530024665116564,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

However the standard in 6.5.2.1 Array subscripting says:
```
If E is an n-dimensional array (n ≥ 2) with dimensions i × j × . . . × k, then E (used as other than an lvalue) is converted to a pointer to an (n − 1)-dimensional array with dimensions j × . . . × k. If the unary * operator is applied to this pointer explicitly, or implicitly as a result of subscripting, the result is the pointed-to (n − 1)-dimensional array, which itself is converted into a pointer if used as other than an lvalue.
```
effectively requiring compatibility between multidimensional array and pointer. Thus it looks like warning should not be generated in such case.

**Comment by Sirraide** [comment]
> However the ISO/IEC 9899:1999 standard in 6.5.2.1 Array subscripting says:
> 
> ```
> If E is an n-dimensional array (n ≥ 2) with dimensions i × j × . . . × k, then E (used as other than an lvalue) is converted to a pointer to an (n − 1)-dimensional array with dimensions j × . . . × k. If the unary * operator is applied to this pointer explicitly, or implicitly as a result of subscripting, the result is the pointed-to (n − 1)-dimensional array, which itself is converted into a pointer if used as other than an lvalue.
> ```
> 
> effectively requiring compatibility between multidimensional array and pointer. Thus it looks like a diagnostics should not be generated in such case as standard compliant compiler should guarantee compatibility between all 3 types `unsigned long *`, `unsigned long [5][6]`, `unsigned long [30]` .

Also, that section in particular is about *subscripting*; the sentence before the part you quoted is
> Successive subscript operators designate an element of a multidimensional array object

so this doesn’t seem like it would apply here given that no subscripting is taking place in the call to `  afunc(mdarr)`.

**Comment by earnol** [comment]
I do not think one needs to be an expert here as we have a standard an it is the one and single source of truth. And my reading tells me that while the whole paragraph indeed talks about subscripting of multidimentional array it (please follow my conjectures):

1. Does establish equivalence `unary * operator is applied to this pointer explicitly, or implicitly as a result of subscripting` between results of the the subscripting.
2. Thus we have an object, which when applied 2 different operations `*` and subscripting that is required to produce identical results for the elements access.
3. Pointer is following the same semantics to provide access to it's element as required equivalency required in the same paragraph.
4. Thus pointers are indeed compatible for the elements access purposes. 
5. There are no standard requirement (or at lest i have not found it) to have those pointers to be incompatible for any other purposes. Thus i conclude it can be implementation defined.
6. From my godbolt experiments i concluded in all compilers the explicit cast produces the compatible pointer for **ALL** valid purposes. Other purposes i checked are address arithmetic and comparisons.
7. The warning currently claims pointer incompatibility which is not correct from both piratical and standard induced (for element access) point-of-views. 
8. Thus the warning or moreover the error is it is spelled now cannot be correct and confusing to the developer and should be addressed.

PS: At the end of the day as Hu Fuming stated: "Practice is the Sole Criterion for Testing Truth" (C)

**Comment by shafik** [comment]
I believe the C FAQ backs up how this currently implemented [My compiler complained when I passed a two-dimensional array to a function expecting a pointer to a pointer.](https://c-faq.com/aryptr/pass2dary.html) in particular:

> ... by which arrays decay into pointers is not applied recursively. ...

**Comment by cor3ntin** [comment]
I think this is inventing language features on the fly.

The standard says (lvalue conversion)

> Except when it is the operand of the sizeof operator, or the _Countof operator, or typeof operators,
or the unary & operator, or is a string literal used to initialize an array, an expression that has type
"array of type" is converted to an expression with type "pointer to type" that points to the initial
element of the array object and is not an lvalue. If the array object has register storage class, the
behavior is implementation-defined.

This cannot be interpreted as "convert array of array of type to pointer to type"

What we might want to consider (if this is really an issue for folks) is to put that case under a different warning flags (`-Wincompatible-array-pointer-types` for example) that can be disabled independently.

Not warning at all on such conversion would require either
  - a small RFC
  - a WG14 paper (we might want to wait for @AaronBallman to come back to weight on the pertinence of such paper)

Note that these specific conversion rules are the same in C++.

**Comment by Endilll** [comment]
> if they did not intended it, they should have explicitly prohibit it in 6.5.2.1 Array subscripting

Both standards first and foremost describe behavior of well-formed programs, in part because the set of ill-formed programs to explicitly prohibit is infinite. One can say this is another spin on "undefined behavior by omission" approach, which is explicitly endorsed by C wording (4.2 Confromance). So you shouldn't draw conclusions from the lack of such prohibition.

**Comment by earnol** [comment]
> So you shouldn't draw conclusions from the lack of such prohibition.

@Endilll Exactly. Lack of specific wording means unknown. If something is not prohibited we cannot say it's use is discouraged. That is why specifically I'm trying to be accurate with my wording and not to make any statements not backed up by the specific wordings in the standard or practical testing results. But in situation when i do have element access equivalence it is direct pointer, that might be compatibility in other areas is desirable as well. In such situations I like to use a duck test: If some unknown object possess some features of another known object, than those two object might be equivalent.  Sometimes this test fails (like with duck imitator of quality), but it is a ground for further clarification and checking.

Anyways as proposed above, let's wait for @AaronBallman and see his position upon demotion of this error back to warning in this specific situation.

**Comment by AaronBallman** [comment]
Thank you for the ping!

Given:
```
int afunc(unsigned long * v);

unsigned long mdarr[5][6];

int main() {
  afunc(mdarr); // #1
}
```
My reading of the standard is that the primary expression `mdarr` at `#1` will decay into a pointer per 6.3.2.1. That converts from `unsigned long[5][6]` to "pointer to type" which points to the initial element of the array object. "Type" in this case is `unsigned long[6]` (6.7.7.3), so that would form the type `unsigned long (*)[6]` which is a pointer to an array object. A pointer to array object is not compatible with a pointer object per the type compatibility rules (6.2.7, 6.7.7.3) and so it's an invalid assignment that violates a constraint (6.5.3.3).

So I think this is all behaving according to spec and changing that would require a paper for WG14. I'm not certain what the committee's appetite for change would be here. One the way to "solve" this would be to change the type compatibility rules so that a pointer to an array of type T and a pointer to type T are compatible. That would then mean this code is accepted as well: https://godbolt.org/z/7sEYan4fT Whether that's a good language design or not is up for debate.
